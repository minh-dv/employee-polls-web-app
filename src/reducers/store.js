import { configureStore } from "@reduxjs/toolkit";
import questionReducer from "./questionSlice";
import userReducer from "./userSlice";
import authReducer from "./authSlice";

export default configureStore({
  reducer: {
    authenticatedUser: authReducer,
    users: userReducer,
    questions: questionReducer,
  },
});
