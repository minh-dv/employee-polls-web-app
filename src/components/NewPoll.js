import React, { useState } from "react";
import { useDispatch } from "react-redux";
import { useNavigate } from "react-router-dom";
import { saveQuestion } from "../util/_DATA";
import { addQuestion } from "../reducers/questionSlice";
import { addQuestionUser } from "../reducers/userSlice";

export const NewPoll = () => {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const [firstOption, setFirstOption] = useState("");
  const [secondOption, setSecondOption] = useState("");

  const loggedInUser = JSON.parse(localStorage.getItem("loggedInUser"));
  const author = loggedInUser.id;
  const handleFirstOptionChange = (e) => {
    setFirstOption(e.target.value);
  };

  const handleSencondOptionChange = (e) => {
    setSecondOption(e.target.value);
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    const question = await saveQuestion(firstOption, secondOption, author);
    if (question) {
      const questionId = question.id;
      dispatch(addQuestion({ questionId, question }));
      dispatch(addQuestionUser({ questionId, author }));
    }
    navigate("/");
  };

  return (
    <div>
      <div className="flex flex-col gap-5 justify-center items-center bg-gray-100 p-5">
        <h2 className="text-2xl font-bold">Poll by sarahedo</h2>
        <div className="h-56 w-56 rounded-full overflow-hidden">
          <img
            className="h-full w-full object-cover"
            src="https://github.com/sarah.png"
            alt={`sarahedo's avatar`}
          />
        </div>
        <h2 className="text-2xl font-bold">Would You Rather</h2>
      </div>

      <div className="max-w-5xl mx-auto my-8 p-8 border border-gray-300 rounded-lg">
        <form onSubmit={handleSubmit}>
          <div className="mb-4">
            <label
              htmlFor="option1"
              className="block font-medium text-gray-700 mb-2"
            >
              First Option
            </label>
            <input
              type="text"
              id="option1"
              name="option1"
              className="w-full border border-gray-300 rounded-lg py-2 px-4"
              placeholder="Option One"
              required
              value={firstOption}
              onChange={handleFirstOptionChange}
            />
          </div>
          <div className="mb-4">
            <label
              htmlFor="option2"
              className="block font-medium text-gray-700 mb-2"
            >
              Second Option
            </label>
            <input
              type="text"
              id="option2"
              name="option2"
              className="w-full border border-gray-300 rounded-lg py-2 px-4"
              placeholder="Option Two"
              required
              value={secondOption}
              onChange={handleSencondOptionChange}
            />
          </div>
          <div className="flex justify-center">
            <button
              type="submit"
              className="bg-green-500 text-white rounded-lg px-4 py-2 hover:bg-green-600 mx-auto w-1/4"
            >
              Create Poll
            </button>
          </div>
        </form>
      </div>
    </div>
  );
};
