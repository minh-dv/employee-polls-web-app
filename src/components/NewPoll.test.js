import { fireEvent, render } from "@testing-library/react";
import store from "../reducers/store";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import React from "react";
import { NewPoll } from "./NewPoll";

describe("NewPoll", () => {
  it("should render the component", () => {
    const user = {
      id: "sarahedo",
      password: "password123",
      name: "Sarah Edo",
      avatarURL: "https://github.com/sarah.png",
      answers: {
        "8xf0y6ziyjabvozdd253nd": "optionOne",
        "6ni6ok3ym7mf1p33lnez": "optionOne",
        am8ehyc8byjqgar0jgpub9: "optionTwo",
        loxhs1bqm25b708cmbf3g: "optionTwo",
      },
      questions: ["8xf0y6ziyjabvozdd253nd", "am8ehyc8byjqgar0jgpub9"],
    };

    localStorage.setItem("loggedInUser", JSON.stringify(user));
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <NewPoll />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });

  it("should display all elements", () => {
    const user = {
      id: "sarahedo",
      password: "password123",
      name: "Sarah Edo",
      avatarURL: "https://github.com/sarah.png",
      answers: {
        "8xf0y6ziyjabvozdd253nd": "optionOne",
        "6ni6ok3ym7mf1p33lnez": "optionOne",
        am8ehyc8byjqgar0jgpub9: "optionTwo",
        loxhs1bqm25b708cmbf3g: "optionTwo",
      },
      questions: ["8xf0y6ziyjabvozdd253nd", "am8ehyc8byjqgar0jgpub9"],
    };

    localStorage.setItem("loggedInUser", JSON.stringify(user));
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <NewPoll />
        </BrowserRouter>
      </Provider>
    );

    const firstOptionLabelElement = component.getByTestId("firstOptionLabel");
    const firstOptionInputElement = component.getByTestId("firstOption");
    const secondOptionLabelElement = component.getByTestId("secondOptionLabel");
    const secondOptionInputElement = component.getByTestId("secondOption");
    const submitButtonElement = component.getByTestId("submit-form");

    expect(firstOptionLabelElement.textContent).toBe("First Option");
    expect(secondOptionLabelElement.textContent).toBe("Second Option");
    expect(submitButtonElement.textContent).toBe("Create Poll");

    fireEvent.change(firstOptionInputElement, {
      target: { value: "FootBall" },
    });
    fireEvent.change(secondOptionInputElement, {
      target: { value: "Voleyball" },
    });
    expect(firstOptionInputElement.value).toBe("FootBall");
    expect(secondOptionInputElement.value).toBe("Voleyball");
  });
});
