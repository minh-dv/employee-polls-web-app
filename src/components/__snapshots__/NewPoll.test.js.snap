// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewPoll should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="flex flex-col gap-5 justify-center items-center bg-gray-100 p-5"
        >
          <h2
            class="text-2xl font-bold"
          >
            Poll by sarahedo
          </h2>
          <div
            class="h-56 w-56 rounded-full overflow-hidden"
          >
            <img
              alt="sarahedo's avatar"
              class="h-full w-full object-cover"
              src="https://github.com/sarah.png"
            />
          </div>
          <h2
            class="text-2xl font-bold"
          >
            Would You Rather
          </h2>
        </div>
        <div
          class="max-w-5xl mx-auto my-8 p-8 border border-gray-300 rounded-lg"
        >
          <form>
            <div
              class="mb-4"
            >
              <label
                class="block font-medium text-gray-700 mb-2"
                data-testid="firstOptionLabel"
                for="option1"
              >
                First Option
              </label>
              <input
                class="w-full border border-gray-300 rounded-lg py-2 px-4"
                data-testid="firstOption"
                id="option1"
                name="option1"
                placeholder="Option One"
                required=""
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-4"
            >
              <label
                class="block font-medium text-gray-700 mb-2"
                data-testid="secondOptionLabel"
                for="option2"
              >
                Second Option
              </label>
              <input
                class="w-full border border-gray-300 rounded-lg py-2 px-4"
                data-testid="secondOption"
                id="option2"
                name="option2"
                placeholder="Option Two"
                required=""
                type="text"
                value=""
              />
            </div>
            <div
              class="flex justify-center"
            >
              <button
                class="bg-green-500 text-white rounded-lg px-4 py-2 hover:bg-green-600 mx-auto w-1/4"
                data-testid="submit-form"
                type="submit"
              >
                Create Poll
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="flex flex-col gap-5 justify-center items-center bg-gray-100 p-5"
      >
        <h2
          class="text-2xl font-bold"
        >
          Poll by sarahedo
        </h2>
        <div
          class="h-56 w-56 rounded-full overflow-hidden"
        >
          <img
            alt="sarahedo's avatar"
            class="h-full w-full object-cover"
            src="https://github.com/sarah.png"
          />
        </div>
        <h2
          class="text-2xl font-bold"
        >
          Would You Rather
        </h2>
      </div>
      <div
        class="max-w-5xl mx-auto my-8 p-8 border border-gray-300 rounded-lg"
      >
        <form>
          <div
            class="mb-4"
          >
            <label
              class="block font-medium text-gray-700 mb-2"
              data-testid="firstOptionLabel"
              for="option1"
            >
              First Option
            </label>
            <input
              class="w-full border border-gray-300 rounded-lg py-2 px-4"
              data-testid="firstOption"
              id="option1"
              name="option1"
              placeholder="Option One"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="mb-4"
          >
            <label
              class="block font-medium text-gray-700 mb-2"
              data-testid="secondOptionLabel"
              for="option2"
            >
              Second Option
            </label>
            <input
              class="w-full border border-gray-300 rounded-lg py-2 px-4"
              data-testid="secondOption"
              id="option2"
              name="option2"
              placeholder="Option Two"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="flex justify-center"
          >
            <button
              class="bg-green-500 text-white rounded-lg px-4 py-2 hover:bg-green-600 mx-auto w-1/4"
              data-testid="submit-form"
              type="submit"
            >
              Create Poll
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
